Simplify dev install using docker

luciancic opened the pull request:
Fixes #560 
----------
## Type of change:
- [ ] Bug fix (change which fixes an issue).
- [ ] New feature (change which adds functionality).
- [ ] Changes to models (requires making migrations).
- [x] Documentation change.
- [x] Architecture change.
----------
## What's Changed:
- Added dockerfile and docker-compose to simplify dev install to 2 commands.
- Updates install docs. Moves current docs as a fallback.
- Note: this is an orthogonal change, it shouldn't affect prod deployments or running the dev env locally as before.

-------
## Checklist:
- [x] My code follows the [style guidelines](https://github.com/LesFruitsDefendus/saskatoon-ng/blob/develop/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.

luciancic commented:
@tristanlatr check it out! 🙂

luciancic commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,20 @@
+    echo "Dependencies not found. Installing..."
+    pip install --no-cache-dir .
+    pip install --no-cache-dir '.[test]'
+    python saskatoon/manage.py migrate
```
Good idea. ✅

luciancic commented on INSTALL.md:
I hesitate to move the docker install docs to a link to the bottom of the original doc. It sounds like bad DX; dev reads and runs whole manual install flow, then arrives at the end and reads "btw you could have done this whole thing in 2 commands". 

Instead I propose an alternative, have the main install.md act as an entry point to the two installation method readmes. I went ahead with that. 

I didn't remove or edit the original instructions in any way, I just moved them to a separate file. I would recommend not tossing aside a good solution only for the sake of making the diff nicer. There are other ways to review a commit with hard-to-read diff, like "View file" or checkout locally.

2ynn commented on saskatoon/saskatoon/settings.py:
```diff
@@ -232,10 +256,14 @@ def get_drf_template_mode() -> str:
 }
 
+# Redis configuration - use Docker service name when in Docker, localhost for local dev
+REDIS_HOST = os.getenv('SASKATOON_REDIS_HOST', '127.0.0.1')
+REDIS_PORT = os.getenv('SASKATOON_REDIS_PORT', '6379')
```
can you please add these two entries to the `.env` template ?

luciancic replied:
Agreed! Good idea

2ynn reviewed (commented):
thanks @luciancic !

Can you please rebase onto the main branch, I think there are diffs unrelated to this PR

2ynn commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,28 @@
+# 1. Check if .env exists; if not, create it from template
+if [ ! -f /app/saskatoon/.env ]; then
+    echo "Creating .env from template"
+    cp /app/saskatoon/env.template /app/saskatoon/.env
```
@luciancic I don't find it very clear what is being used from the `.env` file.  BTW how are you managing it? I think one should be generated instead of copying `SASKATOON_SECRET_KEY='<KEY>'`

luciancic replied:
Ah, quite right. I was copying it in case a dev comes in fresh without having manually copied `env.template` into a `.env` file. But I see now that if I set the secret key in docker-compose.yml then there's no need for any of the other values from the env template.

2ynn commented on docker-compose.yml:
```diff
@@ -0,0 +1,46 @@
+    ports:
+      # if needed you can export a different port for debugging
+      # if the default one on local machine is already used
+      # - "6380:6379"
+      - "6379:6379"
```
shouldn't this match the `SASKATOON_REDIS_PORT` defined in `web.environment` ?

2ynn commented on saskatoon/saskatoon/settings.py:
```diff
@@ -51,6 +52,29 @@
+hostname, _, ips = socket.gethostbyname_ex(socket.gethostname())
+INTERNAL_IPS += [ip[:-1] + '1' for ip in ips]
+# Add common Docker bridge network IPs
+INTERNAL_IPS.extend([
+    '10.0.2.2',
+    '172.17.0.1',
+    '172.18.0.1',
+    '172.19.0.1',
+    '172.20.0.1',
+    '172.21.0.1',
+    '172.22.0.1',
+    '172.23.0.1',
+    '172.24.0.1',
+    '172.25.0.1',
+    '172.26.0.1',
+    '172.27.0.1',
+    '172.28.0.1',
+    '172.29.0.1',
+    '172.30.0.1',
+    '172.31.0.1',
+])
+
```
```suggestion
INTERNAL_IPS.extend(['10.0.2.2'] + [f'172.{i}.0.1' for i in range(17, 32)])
```

luciancic replied:
Nice!

luciancic commented:
Thanks for the feedback @2ynn. I rebased and addressed each point, just one I didn't understand

2ynn commented:
@luciancic I realize I might have been missing the mark w.r.t environment variables in my previous comments. IIUC in your setup, the `.env` file gets loaded in `settings.py` as usual, is that correct? 

What happens then in `settings.py` if my `.env` file contains:
```
SASKATOON_DB_ENGINE=django.db.backends.sqlite3
SASKATOON_DB_NAME=/path/to/db.sqlite3
```
?  Would these take precedence over the ones you are passing to your container: 

```
       - SASKATOON_DB_ENGINE=django.db.backends.mysql
       - SASKATOON_DB_NAME=saskatoon_dev
       - SASKATOON_DB_USER=saskatoon
       - SASKATOON_DB_PASSWORD=saskatoon
       - SASKATOON_DB_HOST=db
```

this is where I got confused

luciancic commented:
Very good points @2ynn!

Indeed the entrypoint script could run pip install, migrations, and fixture on each startup. It adds a bit of overhead to dev startup but I like the simplicity and automatic safe handling of new pip packages, migrations, and fixtures. I tested it by adding my superuser and a new property and re-running the entrypoint, my data was still there (meaning re-running fixtures or migrations doesn't affect existing data as I originally feared).

For env var management I updated the whole approach. If you create a local .env file those values will be used in the web server docker container, if not the values will fallback to defaults hardcoded in docker-compose.yml. Should be the most flexible: new devs can just start it up without any env modification, older devs can customize the env file. The only caveat is that HOST=localhost won't work because docker will look inside its own container, not the local machine. If you set your own env vars you'll have to use HOST=local. I documented this. Afaik there are no other limitations.

2ynn commented on docker-compose.yml:
```diff
@@ -0,0 +1,50 @@
+version: "3.8"
```
I get this warning: `WARN[0000] path/to/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion`

luciancic commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,35 @@
+    echo "Loaded environment variables from .env file"
+else
+    # copy the template file to the .env file, otherwise the app will fail to start
+    cp /app/saskatoon/env.template /app/saskatoon/.env
```
If you have a .env file, the values take precedence over the env vars in docker-compose.yml. I edited the entrypoint script to be a bit clearer and more robust. If no .env file found, copy from template, and then *always load .env* to ensure the execution order (first docker container sets env vars, then entrypointscript sets them from .env file)

For SECRET_KEY I figured it doesn't matter much. The signing just happens with the key "<KEY>" which is not secure for prod but for dev I don't think it matters. Unless I'm missing something.

2ynn replied:
My point is that the fallback values are confusing:
```
     - SASKATOON_SECRET_KEY=secret_key
      - SASKATOON_DB_ENGINE=django.db.backends.mysql
      - SASKATOON_DB_NAME=saskatoon_dev
      - SASKATOON_DB_USER=saskatoon
      - SASKATOON_DB_PASSWORD=saskatoon
      - SASKATOON_DB_HOST=db
      - SASKATOON_DEBUG=yes
      - SASKATOON_REDIS_HOST=redis
      - SASKATOON_REDIS_PORT=6379
```
In which scenario are these values actually used since when I remove the `.env` file it will get recreated from `.env.template` ? Could we simply remove them from `docker-compose.yml`, this way there will be no confusion that the values are set from the `.env`

luciancic replied:
Well in the template.env almost all the values are commented out. So if you do nothing these will get applied. Would it be ok if I add a comment on top of these to say they will get used unless those same ones are defined in .env?

luciancic commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,35 @@
+pip install --no-cache-dir '.[test]'
+
+echo "Running database migrations and initializing fixtures..."
+saskatoon/fixtures/init
```
There's a tradeoff to make a decision on. If we execute it only the first time, then devs will have to manually run fixtures/init when we change them. If we execute every time, yes the values of the fixtures will get reset, _but new records won't be_. That's important to note and I think that makes the tradeoff make more sense in favor of re-running fixtures every time. I'd just leave a note in a limitations section of the readme.

2ynn replied:
>  in favor of re-running fixtures every time

I disagree, I'd rather empower devs and have them understand what the `fixtures/init` script is doing rather than resetting values without warning them. If your goal is to provide a smooth experience for newcomers, running it only on first build is not going against that when they set it up.
In any case we definitely won't want to run it everytime in production, if possible I would prefer working towards a unified docker-compose experience that can eventually be adapted for deployement

2ynn replied:
at the very least I would add a `SASKATOON_INIT_FIXTURES=true` toggle and describe what it does

luciancic commented on doc/setup/docker-install.md:
````diff
@@ -0,0 +1,67 @@
+3. In a new terminal, create a superuser for your site login:
+
+```bash
+docker-compose exec web python saskatoon/manage.py createsuperuser
````
I spent some time going down this route and it's not super-straightforward and I'm not sure it's worth the extra time to save devs a one-time 10 second command. Could we leave that as is for now?

2ynn replied:
nice, thanks!

luciancic commented:
Hey @2ynn I addressed each of your points. For .env I'd argue it's not as needed because new devs can certainly run the docker workflow without having to copy template.env. I'd even say it's a strength of this docker workflow. Once they get to a point they need to edit values in .env, they can and it works seamlessly.

2ynn commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,39 @@
+# Wait for database to be ready and grant privileges needed for tests to work
+echo "Waiting for database to be ready..."
+while ! mysql -h db -u root -proot -e "SELECT 1" &> /dev/null; do
+    echo "Database not ready yet, waiting..."
```
This this dot work for me on fresh install, I get stuck in infinite loop:
```
web-1    | Database not ready yet, waiting...
web-1    | Database not ready yet, waiting...
```

2ynn commented on doc/setup/local-install.md:
```diff
@@ -0,0 +1,231 @@
+# Local Installation guide
```
maybe "manual" is more accurate?

luciancic commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,39 @@
+
+# Wait for database to be ready and grant privileges needed for tests to work
+echo "Waiting for database to be ready..."
+while ! mysql -h db -u root -proot -e "SELECT 1" &> /dev/null; do
```
It did, good catch, fixed

luciancic commented:
Ok, I think I understand your confusion a bit better @2ynn. I moved those default env vars from docker-compose.yml to the template, and added different sections to the template to make it easier to pick up and go. lmkwyt

2ynn commented:
> Ok, I think I understand your confusion a bit better @2ynn. I moved those default env vars from docker-compose.yml to the template, and added different sections to the template to make it easier to pick up and go. lmkwyt

Yes thanks I find this much better! This way there is no possible confusion about some env vars being defined while other values being fallbacks from the compose file. Sorry I wasn't clear right away about what was a possible source of confusion

2ynn commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,41 @@
+fi
+
+{
+    export DJANGO_SUPERUSER_EMAIL=dev@dev.com
+    export DJANGO_SUPERUSER_PASSWORD=dev
```
Would it make sense to also move these to the template file as well and only create a superuser if these env. vars are set ? I'd find it clearer and more flexible

luciancic replied:
Agreed

luciancic commented:
Good feedback @2ynn. We ended up with a much better system than I initially had.
- Single command instead of two ✨
- Clearer install docs
- Single place for all env vars, works out of the box but still configurable ✨

Does this look good to merge?

2ynn commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,41 @@
+    mysql -h "$SASKATOON_DB_HOST" -u root -proot -e "GRANT ALL PRIVILEGES ON *.* TO '$SASKATOON_DB_USER'@'%' WITH GRANT OPTION; FLUSH PRIVILEGES;"
+fi
+
+{
+    # Uses DJANGO_SUPERUSER_EMAIL and DJANGO_SUPERUSER_PASSWORD from .env file
+    python saskatoon/manage.py createsuperuser --noinput
+} || {
+    echo "Failed to create superuser. Possibly because it already exists. Check the logs for more information."
+}
```
I would like the ability to skip calling `createsuperuser` by unsetting those variables:
```suggestion
# set DJANGO_SUPERUSER_EMAIL and DJANGO_SUPERUSER_PASSWORD in .env file to automatically create a superuser
if [ -n "$DJANGO_SUPERUSER_EMAIL" ] && [ -n "$DJANGO_SUPERUSER_PASSWORD" ]; then
    {
        python saskatoon/manage.py createsuperuser --noinput
    } || {
        echo "Failed to create superuser. Possibly because it already exists. Check the logs for more information."
    }
fi
```

2ynn commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,41 @@
+set -e
+
+# Load environment variables from .env file
+if [ ! -f /app/saskatoon/.env ]; then
```
```suggestion
if [ ! -f saskatoon/.env ]; then
```

luciancic commented:
Ok @2ynn I put your suggestions in, and I also updated the host env management. I was finding it tedious to replace localhost with local when switching between containerized and manual setup. So I just had the docker entry point script automatically convert from localhost to local. Fewer sections in the env.template and less things to think about..

2ynn commented on saskatoon/env.template:
```diff
@@ -13,35 +18,84 @@ SASKATOON_DEBUG=yes
+# These settings are for running the web server in Docker
+# You can either:
+# 1. Use the containerized database (DB_HOST=db)
+# 2. Connect to your local database (DB_HOST=localhost or 127.0.0.1)
```
The important part is that I had to grant privileges on `'<user>'@'<docker-ip>'` 

For context, in the the manual installation we give these instructions to create an empty database:
```
    $ sudo mysql -u root -p
    > CREATE USER '<user>'@'localhost' IDENTIFIED BY '<password>';
    > SELECT user FROM mysql.user;   // show all users
    > CREATE DATABASE saskatoon_dev CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
    > SHOW DATABASES;
    > GRANT ALL PRIVILEGES ON saskatoon_dev.* TO '<user>'@'localhost';
```
Note `<user>` here refers to `SASKATOON_DB_USER` defined in `.env`

but this won't work with the current docker setup (option 2) because the connection is not done through `localhost` afaiu

Here are the allowed hosts created by default with the current instructions: 

```
MariaDB [(none)]> SELECT host, user FROM mysql.user;
+------------+----------------------+
| Host       | User                 |
+------------+----------------------+
| localhost  | mariadb.sys |
| localhost  | mysql            |
| localhost  | <user>          |
| localhost  | root               |
+------------+----------------------+
```

In my case I had to make sure that this line was shown: 
```
| 172.18.0.4  | <user>         |
```

Related solution: https://stackoverflow.com/questions/47270505/er-host-not-privileged-docker-container-fails-to-connect-to-mariadb

Here they suggest to create a specific user for the application and grant privileges on all IPs ( i.e. `%` )

luciancic replied:
Ok I added some notes in the readme

2ynn commented:
thanks @luciancic !! 

@tristanlatr do you want to take another look / try it locally or are we good to merge?
